src/OpenTelemetry.Contrib.Instrumentation.AWS/Implementation/AWSTracingPipelineHandler.cs

```diff
@@ -196,10 +196,18 @@ private void AddRequestSpecificInformation(Activity activity, IRequestContext re
             }
         }
 
-        if (AWSServiceHelper.IsDynamoDbService(service))
+        if (AWSServiceType.IsDynamoDbService(service))
         {
             activity.SetTag(SemanticConventions.AttributeDbSystem, AWSSemanticConventions.AttributeValueDynamoDb);
         }
+        else if (AWSServiceType.IsSqsService(service))
+        {
+            Propagators.DefaultTextMapPropagator.Inject(new PropagationContext(activity.Context, Baggage.Current), requestContext, AWSMessagingUtils.SqsMessageAttributeSetter);
```

Comment:
This approach has the problem that we might do a partial injection if the propagator sets multiple keys.

I would suggest to change this to inject into a temporary Dictionary, then check if any of the keys is already contained and if we have enough space in the message attributes to insert all keys. If any of that fails, we shouldn't modify the message attributes ("all-or-nothing").

While the temporary dictionary itself is additional overhead, the two-step-approach can be used to make finding insertion indices faster, since we only need to iterate once to find the highest existing index.

Alternatively, we could also rely on the OriginalRequest to get the list/dictionary of message attributes, maybe that would be easier?

Reply:
- I made some refactoring and now the code for the index determining and key presence relies on the OriginalRequest. 
- Also the code for adding trace data became transactional: "all-or-nothing". 
- I introduced `IRequestContextAdapter` in order to have common code for accessing SNS and SQS attributes properties.